e2e-tests/test/PlutusScripts/BLS.hs

```diff
@@ -377,3 +394,145 @@ verifyBlsGroth16MintWitnessV2 :: C.CardanoEra era
+  }
+
+{-# INLINABLE mkVerifySigG2 #-}
+mkVerifySigG2 :: BlsParamsWithPubKey -> sc -> Bool
+mkVerifySigG2 BlsParamsWithPubKey{..} _sc = do
+  let
+    pkDeser = BI.bls12_381_G2_uncompress pubKey
+    sigDeser = BI.bls12_381_G1_uncompress pubSig
+    hashedMsg = BI.bls12_381_G1_hashToGroup pubMessage blsSigBls12381G2XmdSha256SswuRoNul
+
+  BI.bls12_381_finalVerify (BI.bls12_381_millerLoop pkDeser hashedMsg) (BI.bls12_381_millerLoop g1 sigDeser)
+
+verifyBlsSigG2PolicyV2 :: MintingPolicy
+verifyBlsSigG2PolicyV2 = mkMintingPolicyScript
+  $$(PlutusTx.compile [|| wrap ||])
+  where
+    wrap = mkUntypedMintingPolicy @PlutusV2.ScriptContext mkVerifySigG2
+
+verifyBlsSigG2PolicyScriptV2 :: C.PlutusScript C.PlutusScriptV2
+verifyBlsSigG2PolicyScriptV2 = policyScript verifyBlsSigG2PolicyV2
+
+verifyBlsSigG2AssetIdV2 :: C.AssetId
+verifyBlsSigG2AssetIdV2 = C.AssetId (policyIdV2 verifyBlsSigG2PolicyV2) blsAssetName
+
+verifyBlsSigG2Redeemer :: C.HashableScriptData
+verifyBlsSigG2Redeemer = toScriptData sigG2ParamsWithPubkey
+
+verifyBlsSigG2MintWitnessV2 :: C.CardanoEra era
+  -> (C.PolicyId, C.ScriptWitness C.WitCtxMint era)
+verifyBlsSigG2MintWitnessV2 era =
+    (policyIdV2 verifyBlsSigG2PolicyV2,
+     mintScriptWitness era plutusL2 (Left verifyBlsSigG2PolicyScriptV2) verifyBlsSigG2Redeemer)
+
+---- BLS aggregate signature with same key and different messages with public key over G1 ----
+
+aggregateSigG1Params :: BlsParamsWithAggregateSignature
+aggregateSigG1Params = BlsParamsWithAggregateSignature
+  { aggregateSigMessages =
+    [ "2ba037cdb63cb5a7277dc5d6dc549e4e28a15c70670f0e97787c170485829264"
+    , "ecbf14bddeb68410f423e8849e0ce35c10d20a802bbc3d9a6ca01c386279bf01"
+    , "e8f75f478cb0d159db767341602fa02d3e01c3d9aacf9b686eccf1bb5ff4c8fd"
+    , "21473e89d50f51f9a1ced2390c72ee7e37f15728e61d1fb2c8c839495e489052"
+    , "8c146d00fe2e1caec31b159fc42dcd7e06865c6fa5267c6ca9c5284e651e175a"
+    , "362f469b6e722347de959f76533315542ffa440d37cde8862da3b3331e53b60d"
+    , "73baeb620e63a2e646ea148974350aa337491e5f5fc087cb429173d1eeb74f5a"
+    , "73acc6c3d72b59b8bf5ab58cdcf76aa001689aac938a75b1bb25d77b5382898c"
+    , "4e73ba04bae3a083c8a2109f15b8c4680ae4ba1c70df5b513425349a77e95d3b"
+    , "565825a0227d45068e61eb90aa1a4dc414c0976911a52d46b39f40c5849e5abe"
+    ]
+  , aggregateSigPubKey = "97c919babda8d928d771d107a69adfd85a75cee2cedc4afa4c0a7e902f38b340ea21a701a46df825210dd6942632b46c"
+  , aggregateSignature = "b425291f423235b022cdd038e1a3cbdcc73b5a4470251634abb874c7585a3a05b8ea54ceb93286edb0e9184bf9a852a1138c6dd860e4b756c63dff65c433a6c5aa06834f00ac5a1a1acf6bedc44bd4354f9d36d4f20f66318f39116428fabb88"
+  }
+
+{-# INLINABLE mkAggregateSigG1 #-}
+mkAggregateSigG1 :: BlsParamsWithPubKey -> sc -> Bool
+mkAggregateSigG1 BlsParamsWithPubKey{..} _sc = do
+  let
+    hashedMsgs = mapWithConstSecondArg BI.bls12_381_G2_hashToGroup blsSigBls12381G2XmdSha256SswuRoNul messageElements
```

Comment:
```
blsSigBls12381G2XmdSha256SswuRoNul
```

Excellent variable name!  (I understand why it's like that: it's a pity it's not a bit easier to convert from strings to bytestrings. I wonder if we can use `pack` here to do the conversion off-chain and then lift it into the Plutus code?)

Reply:
I can do
```haskell
blsSigBls12381G2XmdSha256SswuRoNul :: P.BuiltinByteString
blsSigBls12381G2XmdSha256SswuRoNul = BI.toBuiltin $ Char8.pack "BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_"
```
But what's the benefit of lifting rather than simply referencing the variable as I am currently?

Reply:
> But what's the benefit of lifting rather than simply referencing the variable as I am currently?

I forgot about this and now I'm not sure what I was talking about. I think maybe I was alluding to the fact that you need this BuiltinByteString "424c535f5349475f424c53313233383147325f584d443a5348412d3235365f535357555f524f5f4e554c5f" that's supposed to represent a DST containing the ascii string "BLS_blahblahblah" and it's kind of error-prone to have the big hex string rather than a readable string.  I think I was trying to suggest that we could just write the readable string in the source and do the hard work of converting it to a bytestring off the chain (we might be able to convert it on-chain too, but that would surely be too expensive),

Reply:
Oh wait!  I always get horribly confused about his, but isn't `OverloadedStrings` going to make "424c535f..." into the bytestring whose bytes are the ascii codes of the characters '4', '2', '4', 'c', ...?  Don't we in fact want to call some function like `bytesFromHex` as in the Plutus examples?  Maybe that's fixed now though.